activesupport/lib/active_support/testing/instrumentation_assertions.rb

```diff
@@ -0,0 +1,80 @@
+
+# typed: true
+# frozen_string_literal: true
+
+module ActiveSupport
+  module Testing
+    module InstrumentationAssertions
+      # Assert an event was emitted with a given +pattern+ and optional +payload+.
+      #
+      # You can assert that an event was emitted by passing a pattern, which accepts
+      # either a string or regexp, an optional payload, and a block. While the block
+      # is executed, if a matching event is emitted, the assertion will pass.
+      #
+      #     assert_instrumentation_event("post.submitted", title: "Cool Post") do
+      #       post.submit(title: "Cool Post") # => emits matching ActiveSupport::Notifications::Event
+      #     end
+      #
+      def assert_instrumentation_event(pattern, payload = nil, &block)
```

Comment:
I've looked at this PR a few times and I keep getting hung up `instrumentation` being in the class and assertion names. It feels like a mouthful and prone to misspellings. I think something like `assert_event_payload` would be better. I'm curious what other core team members think about the name and if they have other ideas I'm not thinking of.

Reply:
Indeed, `Instrumentation` is bit complex naming wise. In fact, the original naming schema was all based around just `event` such as `assert_event` and `assert_no_events`. We pivoted to the current naming scheme as we thought the former was a bit too generic both semantically and such that it could cause naming conflicts.

As for `Instrumentation`, that was chosen as that's how these events are referred to in the documentation:

> `ActiveSupport::Notifications` provides an instrumentation API for Ruby.

That said, we could revert to `event` alone based naming or perhaps `assert_notification_event` or otherwise? Open to suggestions of course.

Reply:
Ideas from me

- `assert_event` (original)
  - pros: concise, clean
  - bad: low context, naming collision prone
- `assert_instrumentation_event` (current)
  - pros: high context, low chance of collision
  - cons: long, harder to say/spell, perhaps too verbose
- `assert_notifications_event`
  - pros: relatively high context, less hard to spell/say than `instrumentation`
  - cons: long, could use `notification` I suppose
- `assert_instrumented_event` or `assert_subscribed_event`
  - pros: relatively high context
  - cons: semantically feels like it should have `emitted` at the end or something similar

as well as from Matthew D

> IMO: Instrument is the verb; what it produces is notifications
> `assert_notification` / `assert_no_notifications`

and Christian Bica

> As `ActiveSupport::Notifications` emits events I'd use `assert_emits_event`

Reply:
:wave: @eileencodes, when you have a moment, please let me know what you think of the new naming scheme and any other feedback. I think it's in a pretty good state now and would like to get it merged for 8.0 if possible. :pray: 

Can also use this new module to cleanup much of the [existing rails/rails codebase](https://github.com/rails/rails/pull/53065#discussion_r1780342475) once merged. :slightly_smiling_face: